DeeperFunc/Router.swift

```diff
@@ -0,0 +1,206 @@
+        self.scheme = scheme
+        self.rootDeepLinkHandler = rootDeepLinkHandler
+    }
+
+    public var description: String {
+        return "\(type(of: self)):\n[\(route?.template ?? "")]"
+    }
+    
+    public func openURL(_ url: URL) -> U? {
+        guard
+            let components = url.routeComponents,
+            url.scheme == scheme,
+            let result = route?.parse(components),
+            result.rest.path.isEmpty
+            else { return nil }
+        
+        return result.match
+    }
+
+    private func add(_ route: RoutePattern<U, Path>) -> Router {
+        self.route = self.route.map({ oldValue in
+            .init(parse: { oldValue.parse($0) ?? route.parse($0) },
+                  print: { oldValue.print($0) ?? route.print($0) },
+                  template: "\(oldValue.template)\n\(route.template)")
+        }) ?? route
+        return self
+    }
+    
+    @discardableResult
+    public func add(_ intent: U, route: String) -> Router {
+        return add(intent, route: lit(route))
+    }
+    
+    @discardableResult
+    public func add<S: ClosedPatternState>(_ intent: U, route: RoutePattern<Void, S>) -> Router {
+        return add(route.map({ intent }, { $0 == intent ? () : nil }))
+    }
+    
+    @discardableResult
+    public func add(_ intent: U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        let typedRoute: RoutePattern<Void, Path> = route.map({ _ in () }, { _ in () })
+        return add(intent, route: typedRoute)
+    }
+
+    @discardableResult
+    public func add<A, S: ClosedPatternState>(_ intent: @escaping ((A)) -> U, route: RoutePattern<A, S>) -> Router {
+        return add(route.map({ intent($0) }, { $0.deconstruct(intent) }))
+    }
+
+    @discardableResult
+    public func add<A>(_ intent: @escaping ((A)) -> U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        let typedRoute: RoutePattern<A, Path> = route.map({ $0 as? A }, { $0 })
+        return add(intent, route: typedRoute)
+    }
+    
+    @discardableResult
+    public func add<A, B, S: ClosedPatternState>(_ intent: @escaping ((A, B)) -> U, route: RoutePattern<(A, B), S>) -> Router {
+        return add(route.map({ intent($0) }, { $0.deconstruct(intent) }))
+    }
+
+    @discardableResult
+    public func add<A, B>(_ intent: @escaping ((A, B)) -> U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        let typedRoute: RoutePattern<(A, B), Path> = route.map({ $0 as? (A, B) }, { $0 })
+        return add(intent, route: typedRoute)
+    }
+
+    @discardableResult
+    public func add<A, B, C, S: ClosedPatternState>(_ intent: @escaping ((A, B, C)) -> U, route: RoutePattern<((A, B), C), S>) -> Router {
+        return add(route.map({ intent(flatten($0)) }, { $0.deconstruct(intent).map(parenthesize) }))
```

Comment:
Here we also need to regroup values from `(A, B, C)` to `((A, B), C)` as this is what route expects